crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs

```diff
@@ -0,0 +1,566 @@
+    /// When two branches are merged back into one, this consitutes a join point, and is analogous
+    /// to the rest of the program after an if statement. When such a join point / end block is
+    /// found, the top of this conditions stack is popped since we are no longer under that
+    /// condition. If we are under multiple conditions (a nested if), the topmost condition is
+    /// the most recent condition combined with all previous condtions via `And` instructions.
+    conditions: Vec<(BasicBlockId, ValueId)>,
+
+    /// A map of values from the unmodified function to their values given from this pass.
+    /// In particular, this pass will remove all block arguments except for function parameters.
+    /// Each value in the function's entry block is also left unchanged.
+    values: HashMap<ValueId, ValueId>,
+}
+
+fn flatten_function_cfg(function: &mut Function) {
+    let mut context = Context {
+        cfg: ControlFlowGraph::with_function(function),
+        function,
+        branch_ends: HashMap::new(),
+        conditions: Vec::new(),
+        values: HashMap::new(),
+    };
+    context.flatten();
+}
+
+impl<'f> Context<'f> {
+    fn flatten(&mut self) {
+        self.analyze_function();
+
+        // Start with following the terminator of the entry block since we don't
+        // need to flatten the entry block into itself.
+        self.handle_terminator(self.function.entry_block());
+    }
+
+    /// Visits every block in the current function to find all blocks with a jmpif instruction and
+    /// all blocks which terminate the jmpif by having each of its branches as a predecessor.
+    fn analyze_function(&mut self) {
+        let post_order = PostOrder::with_function(self.function);
+        let dom_tree = DominatorTree::with_cfg_and_post_order(&self.cfg, &post_order);
+        let mut branch_beginnings = Vec::new();
+
+        let mut visited = HashSet::new();
+        let mut queue = VecDeque::new();
+        queue.push_front(self.function.entry_block());
+        visited.insert(self.function.entry_block());
+
+        while let Some(block_id) = queue.pop_front() {
+            // If there is more than one predecessor, this must be an end block
+            let mut predecessors = self.cfg.predecessors(block_id);
+            if predecessors.len() > 1 {
+                // If we haven't already visited all of this block's predecessors, delay analyzing
+                // the block until we have. This ensures we analyze the function in evaluation order.
+                if !predecessors.all(|block| visited.contains(&block)) {
+                    queue.push_back(block_id);
+                    continue;
+                }
+
+                // We expect the merging of two branches to be ordered such that only the most
+                // recent jmpif is a candidate for being the start of the two branches merged by
+                // a block with 2 predecessors.
```

Comment:
Good catch, and fixed. This was an issue with a change I made to append visited blocks via `visited.extend(block.successors())` before they're actually visited, in order to prevent blocks from being added twice to the queue. I've removed this and reverted back to marking a block visited when we actually visit it and just adding a check to test if it was visited twice. Good test case as well.